File: filter-event-and-function-py/src/agent.py
from .large_transfer_event import handle_transaction as handle_large_transfer_event
from .transfer_from_function import handle_transaction as handle_transfer_from_function

findings_count = 0


def provide_handle_transaction(handle_large_transfer_event, handle_transer_from_function):
    def handle_transaction(transaction_event):
        # limiting this agent to emit only 5 findings so that the alert feed is not spammed
        global findings_count
        if findings_count >= 5:
            return []

        findings = handle_large_transfer_event(
            transaction_event) + handle_transer_from_function(transaction_event)

        findings_count += len(findings)
        return findings

    return handle_transaction


real_handle_transaction = provide_handle_transaction(
    handle_large_transfer_event, handle_transfer_from_function)


def handle_transaction(transaction_event):
    return real_handle_transaction(transaction_event)

File: adding-labels/src/agent.ts
import {
  Finding,
  HandleTransaction,
  TransactionEvent,
  FindingSeverity,
  FindingType,
  EntityType,
} from "forta-agent";

import {
  ERC20_TRANSFER_EVENT,
  TETHER_ADDRESS,
  TETHER_DECIMALS,
} from "./constants";
let findingsCount = 0;

const handleTransaction: HandleTransaction = async (
  txEvent: TransactionEvent
) => {
  const findings: Finding[] = [];

  // limiting this agent to emit only 5 findings so that the alert feed is not spammed
  if (findingsCount >= 5) return findings;

  // filter the transaction logs for Tether transfer events
  const tetherTransferEvents = txEvent.filterLog(
    ERC20_TRANSFER_EVENT,
    TETHER_ADDRESS
  );

  tetherTransferEvents.forEach((transferEvent) => {
    // extract transfer event arguments
    const { to, from, value } = transferEvent.args;
    // shift decimals of transfer value
    const normalizedValue = value.div(10 ** TETHER_DECIMALS);

    // if more than 10,000 Tether were transferred, report it
    if (normalizedValue.gt(10000)) {
      findings.push(
        Finding.fromObject({
          name: "High Tether Transfer",
          description: `High amount of USDT transferred: ${normalizedValue}`,
          alertId: "FORTA-1",
          severity: FindingSeverity.High,
          type: FindingType.Suspicious,
          metadata: {
            from,
            to,
          },
          labels: [
            {
              entity: from,
              entityType: EntityType.Address,
              label: "attacker",
              confidence: 0.6,
              metadata: {},
              remove: false,
            },
          ],
        })
      );
      findingsCount++;
    }
  });

  return findings;
};

export default {
  handleTransaction,
};
